src/ast/passes/semantic_analyser.cpp

```diff
@@ -1715,6 +1715,46 @@ void SemanticAnalyser::validate_map_key(const SizedType &key, Node &node)
   }
 }
 
+void SemanticAnalyser::visit(MapDeclStatement &decl)
+{
+  if (!bpftrace_.config_->get(ConfigKeyBool::unstable_map_decl)) {
+    decl.addError() << "Map declarations are not enabled by default. To enable "
+                       "this unstable feature, set this config flag to 1 "
+                       "e.g. unstable_map_decl=1";
+  }
+
+  auto found = BPF_MAP_TYPES.find(decl.bpf_type);
+  if (found == BPF_MAP_TYPES.end()) {
+    auto &err = decl.addError();
+    err << "Invalid bpf map type: " << decl.bpf_type;
+    auto &hint = err.addHint();
+    hint << "Valid map types: ";
+    int num_types = BPF_MAP_TYPES.size();
+    for (const auto &bpf_type : BPF_MAP_TYPES) {
+      hint << bpf_type.first;
+      num_types--;
+      if (num_types != 0) {
+        hint << ", ";
+      }
+    }
+  } else {
+    bpf_map_type_.insert({ decl.ident, found->second });
+
+    if (decl.max_entries != 1 &&
+        found->second == libbpf::BPF_MAP_TYPE_PERCPU_ARRAY) {
+      decl.addError() << "Max entries can only be 1 for map type "
+                      << decl.bpf_type;
+    }
+  }
+
+  if (!is_first_pass()) {
```

Comment:
Since warnings won't block progress, I think maybe this should be `is_final_pass` instead of `!is_first_pass()`? Otherwise, I think this warning could be added multiple times.